src/connection/host_lookup.c

```diff
@@ -0,0 +1,694 @@
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * host_lookup.c
+ */
+
+#ident "$Id$"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <string.h>
+#include <limits.h>
+#include <errno.h>
+#include <unordered_map>
+
+#if defined (WINDOWS)
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#include <io.h>
+#else
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <pthread.h>
+#endif
+
+#include "porting.h"
+#include "system_parameter.h"
+#include "environment_variable.h"
+#include "message_catalog.h"
+
+#define LINE_BUF_SIZE                (512)
+#define HOSTNAME_LEN                 (256)
+#define MAX_NUM_HOSTS                (256)
+#define IPADDR_LEN                   (17)
+#define MTIME_DIGITS                 (10)
+#define IPv4_ADDR_LEN                (4)
+#define NUM_IPADDR_DOT               (3)
+#define MAX_NUM_IPADDR_PER_HOST      (1)
+#define USER_HOSTS_FILE              "cubrid_hosts.conf"
+
+#define NUM_DIGIT(VAL)              (size_t)(log10 (VAL) + 1)
+#define FREE_MEM(PTR)           \
+        do {                    \
+          if (PTR) {            \
+            free(PTR);          \
+            PTR = 0;            \
+          }                     \
+        } while (0)
+
+typedef enum
+{
+  HOSTNAME_TO_IPADDR = 0,
+  IPADDR_TO_HOSTNAME = 1,
+} LOOKUP_TYPE;
+
+typedef enum
+{
+  INSERT_IPADDR = 0,
+  INSERT_HOSTNAME = 1,
+} HOSTENT_INSERT_TYPE;
+
+typedef enum
+{
+  USE_GLIBC_HOSTS = 0,
+  USE_USER_DEFINED_HOSTS = 1,
+} HOST_LOOKUP_TYPE;
+
+typedef enum
+{
+  LOAD_FAIL = -1,
+  LOAD_INIT,
+  LOAD_SUCCESS,
+} HOSTS_CONF_LOAD_STATUS;
+
+static struct hostent *hostent_Cache[MAX_NUM_HOSTS];
+
+static int hosts_conf_file_Load = LOAD_INIT;
+
+static pthread_mutex_t load_hosts_file_lock = PTHREAD_MUTEX_INITIALIZER;
+
+// *INDENT-OFF*
+static std::unordered_map <std::string, int> user_host_Map;
+// *INDENT-ON*
+
+static struct hostent *hostent_alloc (char *ipaddr, char *hostname);
+static bool ip_format_check (char *ip_addr);
+static int load_hosts_file ();
+static struct hostent *host_lookup_internal (const char *hostname, struct sockaddr *saddr, LOOKUP_TYPE lookup_type);
+
+/*
+ * hostent_alloc () - Allocate memory hostent structure.
+ * 
+ * return        : The hostent pointer.
+ * ipaddr (in)   : Elements of hostent.
+ * hostname (in) : Elements of hostent.
+ */
+static struct hostent *
+hostent_alloc (char *ipaddr, char *hostname)
+{
+  struct hostent *hp;
+  char addr_trans_bi_buf[sizeof (struct in_addr)];
+
+  if ((hp = (struct hostent *) malloc (sizeof (struct hostent))) == NULL)
+    {
+      return NULL;
+    }
+
+  hp->h_addrtype = AF_INET;
+  hp->h_length = IPv4_ADDR_LEN;
+
+  if (inet_pton (AF_INET, ipaddr, addr_trans_bi_buf) < 1)
+    {
+      FREE_MEM (hp);
+      goto return_phase;
+    }
+
+  hp->h_name = strdup (hostname);
+  hp->h_aliases = NULL;
+
+  if ((hp->h_addr_list = (char **) malloc (sizeof (char *) * MAX_NUM_IPADDR_PER_HOST)) == NULL)
```

Comment:
Is there any reason to use sizeof(char *)?

Reply:
There is no special reason but I knew it is common usage for malloc.
"dynamic_load.c" +586 also uses this format.